Implement backend unit tests

GenFusion122 opened the pull request:
Implemented unit tests on each backend service using Jasmine.
Each individual microservice has its own spec folder containing .spec files used for testing.
To run a test use the command:
`npx jasmine --random=false SPEC_FILE` 
e.g. `npx jasmine --random=false .\socket.spec.js`

Note: Due to the way matching service works in delaying a response, it may fail occasionally. In this event, just rerun the test and it should pass.

GenFusion122 committed:
Implement unit tests

Implemented unit tests on each backend service using Jasmine.

chia-yh commented on backend/collab-service/spec/socket.spec.js:
```diff
@@ -0,0 +1,64 @@
+const { createServer } = require("node:http");
```
Should there be unit tests testing the service's response to certain events? It seems to just be testing the socket.io instead of any specific functionality of the service

GenFusion122 replied:
That's the issue with socket io and unit testing, I was not able to find an equivalent library to supertest which is what the others used to "run" the index.js essentially.

chia-yh replied:
In that case I don't see the point of adding these test cases, as they essentially test just the library

GenFusion122 replied:
Yeah I suppose that's fair. I'll remove along with the other changes requested.

chia-yh commented on backend/matching-service/index.js:
```diff
@@ -33,4 +33,6 @@ require('./routes/matching.routes')(app);
\ No newline at end of file
+});
+
+module.exports = app;
```
Is this necessary?

GenFusion122 replied:
This is required to call the app in the unit tests.

chia-yh commented on backend/question-service/index.js:
```diff
@@ -27,17 +27,33 @@ const PORT = 8002;
+const { MongoMemoryServer } = require('mongodb-memory-server');
 
-db.mongoose
+if (process.env.JASMINE) {
```
Should this be in the implementation? It bears no relevance outside of testing, my opinion is that it should be elsewhere

GenFusion122 replied:
Since the unit tests use the index.js file, its needed to determine if its running in a test instance or not. If it is then it runs the database as a mock using the `mongodb-memory-server`.

chia-yh replied:
On principle anything related to test code, especially something explicit like this, should not be present in any code for deployment. I would favour the removal of this and see if there's a manner in which you can set up a mock to listen for a connection via another set of configs set through an environment file.

chia-yh replied:
Can I check what requests are being tested with the stub

GenFusion122 replied:
The stub is basically a mock of the database since these tests are to be run outside the container, all tests involving any controller with a database is run with the stub.

As for the environmental variables, would it be better to have a "testing copy" of the index.js to serve the same purpose? Considering all the index.js files are setup in the same manner?

chia-yh commented on frontend/src/app/matching/matching.component.ts:
```diff
@@ -45,7 +45,7 @@ export class MatchingComponent {
 
   cancelMatch() {
-    let userid = this.storageService.getUser()["id"]
+    let userid = this.storageService.getUser()["username"]
```
Change should be made elsewhere if relevant to functionality of deployed software

GenFusion122 replied:
Yeah this was a change I made to fix the id thing I mentioned, its probably better to be included in the bug fixes with @cjyothika's current PR. I'll remove it from here and let her know.

chia-yh commented on backend/matching-service/spec/matching.spec.js:
```diff
@@ -0,0 +1,113 @@
+  let match1;
+  let match2;
+
+  beforeAll(() => {
```
Might want to ensure a queue is empty before each test case as well, to prevent side-effects.

Also, as mentioned in the PR description, which tests are failing as a result of the nature of the service?

GenFusion122 replied:
Its the "should enqueue a user" test that's failing, its primarily because of the way we implemented the request having no response until a match is found or it times out. At least I think that's the most likely reason for failure, but it passes the case a vast majority of the time. We could leave out this matching test case since we already have unit tests for the queue itself?

GenFusion122 replied:
Updated with a timeout, works fine now.

chia-yh reviewed (changes requested):
Some rather major issues that should be addressed. If the unit tests aren't complete or have major issues might need to consider not implementing them as they may not accurately test desired functionality which could lead to problems

GenFusion122 committed:
Implement more modular unit tests and minor improvements

All tests now use `app.spec.js` to run the server as opposed to running their respective index.js.

Added a timeout of 1 second to unit tests for matching service, this should fix the issue of it failing occasionally.

Improved some tests for question service.

chia-yh reviewed (approved):
LGTM
